### PR TITLE
oop-object_independence-hard-q2

### PR DESCRIPTION
**Pull request by SamamthaH**
## PR Checklist
- [x]  Are helper methods really needed here?
    > Opt for hardcoding input/output into the input provider when possible
- [x]  Are the assertion fail messages good?
    - [x]  Logical (does it make sense?)
    - [x]  Readable (can a student understand what’s wrong?)
    - [x]  Not too feed-y (does it give away the answer?)
    - [x]  Grammatically sound (is the grammar and punctuation in the sentence proper?)
    - [x]  Specific (does the fail message relate to the question?)
- [x]  Is the code formatted well?
    > Run `CTRL` + `ALT` + `L` in IntelliJ
- [x]  EOF newlines
- [x]  Do the comments for Parsons follow the correct convention?
- [x]  Is the question being tested sufficiently?
    - [x]  Are there representative cases to test the program?
    - [x]  What about edge cases? (empty strings, negatives, infinity, off-by-one errors, etc)
        > If the question doesn’t specify what to do, mark it as a bad question
- [x]  Are the test method names good?
    > Test method names show up in a “What went well” section on the website, so “mathTest” will become “What went well: math test”, which doesn’t make much sense. A better name in this example would be “calculatesTaxCorrectly” so it shows up as “What went well: calculates tax correctly”

### CODE REVIEW

File: src/oop/object_independence/hard/q2/MainTest.java [hunk 1]
```diff
@@ -0,0 +1,237 @@
...
+                Arguments.of("Red", "Nb", 50),
+                Arguments.of("Magenta", "NA", 1738274),
+                Arguments.of("Neon Blue", "Female", 1),
+                Arguments.of("Pinkest Pink", "Male", 164)
+        );
+    }
+
+    @ParameterizedTest
+    @MethodSource("constructorInputProvider")
+    public void courseConstructorInitializesFieldsCorrectly(String c, String g, int a) throws Throwable {
```
Comment (SamamthaH) [inline src/oop/object_independence/hard/q2/MainTest.java:88]:
```suggestion
    public void fishConstructorInitializesFieldsCorrectly(String c, String g, int a) throws Throwable {
```

File: src/oop/object_independence/hard/q2/MainTest.java [hunk 2]
```diff
@@ -0,0 +1,237 @@
...
+                Arguments.of("Pink", "Xeno", 2, "Brown", "Male", 198),
+                Arguments.of("Cosmic Latte", "Male", 0, "Scarlet", "Female", 32),
+                Arguments.of("Amethyst", "NA", 16, "Purple", "Male", 2),
+                Arguments.of("Orange", "Male", 678, "Yellow", "NA", 0)
+        );
+    }
+
+    @ParameterizedTest
+    @MethodSource("setMethodsInputProvider")
+    public void platformGameClassHasCorrectSetNumLevelsMethod(String colour, String gender, int age, String newColour, String newGender, int newAge) throws Throwable {
```
Comment (SamamthaH) [inline src/oop/object_independence/hard/q2/MainTest.java:139]:
```suggestion
    public void fishHasCorrectSetMethods(String colour, String gender, int age, String newColour, String newGender, int newAge) throws Throwable {
```

File: src/oop/object_independence/hard/q2/MainTest.java [hunk 3]
```diff
@@ -0,0 +1,237 @@
...
+                String.format(incorrectSetGet, getStringAttributeName2, stringAttributeName2, setStringAttributeName2));
+        _assertEquals(newAge, output3,
+                String.format(incorrectSetGet, getIntAttributeName1, intAttributeName1, setIntAttributeName1));
+        _assertEquals(expectedToStringOutput, toStringOutput,
+                String.format(incorrectToString, stringAttributeName1, stringAttributeName2, intAttributeName1,
+                        setStringAttributeName1, setStringAttributeName2, setIntAttributeName1));
+    }
+
+    @Test
+    public void testPlatformGameMainMethodProducesCorrectOutput() throws Throwable {
```
Comment (SamamthaH) [inline src/oop/object_independence/hard/q2/MainTest.java:211]:
```suggestion
    public void testAquariumMainMethodProducesCorrectOutput() throws Throwable {
```

File: src/oop/object_independence/hard/q2/MainTest.java [hunk 4]
```diff
@@ -0,0 +1,237 @@
...
+                        testStringAttributeName1a, testStringAttributeName1b, testIntAttributeName1 + testIntAttributeNameUpdate)),
+                new NewLine(),
+                new StringLiteral(String.format("This fish is %s, and is %s, and is %d years old",
+                        testStringAttributeName2a, testStringAttributeName2b, testIntAttributeName2 + testIntAttributeNameUpdate)),
+                new NewLine(),
+                new StringLiteral(String.format("This fish is %s, and is %s, and is %d years old",
+                        testStringAttributeName3a, testStringAttributeName3b, testIntAttributeName3 + testIntAttributeNameUpdate)),
+        };
+        String incorrectOutput =
+                String.format("Your %s class main method does correctly initialize, update, and print the values of the two %s objects.",
```
Comment (SamamthaH) [inline src/oop/object_independence/hard/q2/MainTest.java:233]:
```suggestion
                String.format("Your %s class main method does not correctly initialize, update, and print the values of the two %s objects.",
```

File: src/oop/object_independence/hard/q2/MainTest.java [hunk 5]
```diff
@@ -0,0 +1,237 @@
...
+                        testStringAttributeName1a, testStringAttributeName1b, testIntAttributeName1 + testIntAttributeNameUpdate)),
+                new NewLine(),
+                new StringLiteral(String.format("This fish is %s, and is %s, and is %d years old",
+                        testStringAttributeName2a, testStringAttributeName2b, testIntAttributeName2 + testIntAttributeNameUpdate)),
+                new NewLine(),
+                new StringLiteral(String.format("This fish is %s, and is %s, and is %d years old",
+                        testStringAttributeName3a, testStringAttributeName3b, testIntAttributeName3 + testIntAttributeNameUpdate)),
+        };
+        String incorrectOutput =
+                String.format("Your %s class main method does not correctly initialize, update, and print the values of the two %s objects.",
```
Comment (SamamthaH) [inline src/oop/object_independence/hard/q2/MainTest.java:233]:
```suggestion
                String.format("Your %s class main method does not correctly initialize, update, and print the values of the three %s objects.",
```

File: src/oop/object_independence/hard/q2/MainTest.java [hunk 6]
```diff
@@ -0,0 +1,237 @@
...
+                String.format(incorrectVisibilityModifier, objectClassName, stringAttributeName2));
+        assertTrue(classInstance.hasModifier(intAttributeName1, "private"),
+                String.format(incorrectVisibilityModifier, objectClassName, intAttributeName1));
+    }
+
+    @Test
+    public void fishClassHasRequiredConstructor() {
+        String missingConstructorMessage = "Your " + testClassName + " class is missing a required constructor.";
+        Class<?>[] arguments = {String.class, String.class, int.class};
+        assertTrue(classInstance.hasConstructor(arguments), missingConstructorMessage);
```
Comment (KaiKhademi) [inline src/oop/object_independence/hard/q2/MainTest.java:74]: Check the constructor modifier here as well.

File: src/oop/object_independence/hard/q2/MainTest.java [hunk 7]
```diff
@@ -0,0 +1,237 @@
...
+    @ParameterizedTest
+    @MethodSource("constructorInputProvider")
+    public void fishConstructorInitializesAttributesCorrectly(String c, String g, int a) throws Throwable {
+        Object[][] arguments = {
+                {c, String.class},
+                {g, String.class},
+                {a, int.class}
+        };
+        String incorrectVisibilityModifier = "Your %s constructor has the incorrect visibility modifier.";
+        assertTrue(classInstance.hasModifier(new Class[]{String.class, String.class, int.class}, "public"),
```
Comment (KaiKhademi) [inline src/oop/object_independence/hard/q2/MainTest.java:95]: I think it's better it this is in the hasConstructor test

File: src/oop/object_independence/hard/q2/MainTest.java [hunk 8]
```diff
@@ -0,0 +1,237 @@
...
+                Arguments.of("Red", "Female", 18, "Pansy", "NA", 28),
+                Arguments.of("Pink", "Xeno", 2, "Brown", "Male", 198),
+                Arguments.of("Cosmic Latte", "Male", 0, "Scarlet", "Female", 32),
+                Arguments.of("Amethyst", "NA", 16, "Purple", "Male", 2),
+                Arguments.of("Orange", "Male", 678, "Yellow", "NA", 0)
+        );
+    }
+
+    @ParameterizedTest
+    @MethodSource("setMethodsInputProvider")
```
Comment (KaiKhademi) [inline src/oop/object_independence/hard/q2/MainTest.java:138]: We should make different test for each method here both for accessors and mutators. Because if they write one of the corectly they should get the test for that.

**Review by KaiKhademi — CHANGES_REQUESTED** [review]

File: src/oop/object_independence/hard/q2/MainTest.java [hunk 9]
```diff
@@ -0,0 +1,292 @@
...
+        assertTrue(classInstance.hasModifier(intAttributeName1, "private"),
+                String.format(incorrectVisibilityModifier, objectClassName, intAttributeName1));
+    }
+
+    @Test
+    public void fishClassHasRequiredConstructor() {
+        Class<?>[] arguments = {String.class, String.class, int.class};
+        assertTrue(classInstance.hasConstructor(arguments),
+                String.format("Your %s class is missing a required constructor.", testClassName));
+        assertTrue(classInstance.hasConstructor(arguments, new String[]{"public"}),
```
Comment (KaiKhademi) [inline src/oop/object_independence/hard/q2/MainTest.java:75]: Refer to discord.

**Review by KaiKhademi — CHANGES_REQUESTED** [review]

**Review by KaiKhademi — DISMISSED** [review]

File: src/oop/object_independence/hard/q2/QuestionText.md [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+Your `TestAquarium` class must have the following items:
+
+- Main Method:
+    - The main method should initialize three `Fish` objects
+    - The first object should have a **colour** of `Blue`, a **gender** of `Male`, and an **age** of `2`
+    - The second object should have **colour** of `White`, a **gender** of `Female`, and an **age** of `4`
+    - The third object should have a **colour** of `Red`, a **gender** of `Female`, and an **age** of `3`
+    - After initializing both objects, print the `toString` value of each object, starting with the first object
+    - Next, using the `setAge` method, update the **age** attribute of all objects to be 10 higher
+    - Finally, print the `toString` of both objects again.
```
Comment (Hedgemon4) [inline src/oop/object_independence/hard/q2/QuestionText.md:50]:
```suggestion
    - Finally, print the `toString` of all the objects again.
```

File: src/oop/object_independence/hard/q2/MainTest.java [hunk 11]
```diff
@@ -0,0 +1,299 @@
...
+        String objectClassString = "oop.object_independence.hard.q2." + objectClassName;
+        String testClassString = "oop.object_independence.hard.q2." + testClassName;
+        classInstance = new ObjectTest(objectClassString);
+        testInstance = new ObjectTest(testClassString);
+    }
+
+    @Test
+    public void fishClassHasCorrectAttributes() {
+        String missingAttributeMessage = "Your %s class is missing the %s attribute. Please make sure you have added it and that it is spelled correctly.";
+        assertTrue(classInstance.hasField(stringAttributeName1, String.class),
```
Comment (Hedgemon4) [inline src/oop/object_independence/hard/q2/MainTest.java:55]: This `hasField` call is checking not just the name, but also the class already. Please check them separately.

File: src/oop/object_independence/hard/q2/MainTest.java [hunk 12]
```diff
@@ -0,0 +1,299 @@
...
+                String.format(incorrectVisibilityModifier, objectClassName, stringAttributeName2));
+        assertTrue(classInstance.hasModifier(intAttributeName1, "private"),
+                String.format(incorrectVisibilityModifier, objectClassName, intAttributeName1));
+    }
+
+    @Test
+    public void fishClassHasRequiredConstructor() {
+        Class<?>[] arguments = {String.class, String.class, int.class};
+        assertTrue(classInstance.hasConstructor(arguments),
+                String.format("Your %s class is missing a required constructor.", testClassName));
```
Comment (Hedgemon4) [inline src/oop/object_independence/hard/q2/MainTest.java:81]:
```suggestion
                String.format("Your %s class is missing a required constructor. Please check you have declared the constructor and it has the correct parameters.", testClassName));
```

File: src/oop/object_independence/hard/q2/MainTest.java [hunk 13]
```diff
@@ -0,0 +1,299 @@
...
+    @MethodSource("constructorInputProvider")
+    public void fishHasCorrectGetColourMethod(String colour, String gender, int age) throws Throwable {
+        Object[][] arguments = {
+                {colour, String.class},
+                {gender, String.class},
+                {age, int.class}
+        };
+        Object instance = classInstance.createInstance(arguments);
+        String incorrectGetMethods = "Your %s method does not return the value of the %s attribute.";
+
```
Comment (Hedgemon4) [inline src/oop/object_independence/hard/q2/MainTest.java:124]: Please check the method definition first

File: src/oop/object_independence/hard/q2/MainTest.java [hunk 14]
```diff
@@ -0,0 +1,299 @@
...
+        _assertEquals(newGender, output2,
+                String.format(incorrectSetGet, getStringAttributeName2, stringAttributeName2, setStringAttributeName2));
+        _assertEquals(newAge, output3,
+                String.format(incorrectSetGet, getIntAttributeName1, intAttributeName1, setIntAttributeName1));
+        _assertEquals(expectedToStringOutput, toStringOutput,
+                String.format(incorrectToString, stringAttributeName1, stringAttributeName2, intAttributeName1,
+                        setStringAttributeName1, setStringAttributeName2, setIntAttributeName1));
+    }
+
+    @Test
```
Comment (Hedgemon4) [inline src/oop/object_independence/hard/q2/MainTest.java:272]: This test should be dependent on others using tags. Please see https://github.com/canvas-gamification/canvas-gamification-junit-tests/pull/694 for an example.

File: src/oop/object_independence/hard/q2/MainTest.java (hunk 13 above)
Comment (Hedgemon4) [reply src/oop/object_independence/hard/q2/MainTest.java:124]: This applies to all the tests for the methods below

**Review by Hedgemon4 — CHANGES_REQUESTED** [review]

File: src/oop/object_independence/hard/q2/MainTest.java [hunk 15]
```diff
@@ -0,0 +1,343 @@
...
+
+        _assertEquals(newAge, classInstance.getFieldValue(instance, intAttributeName1),
+                String.format(incorrectSetMethods, setIntAttributeName1, intAttributeName1));
+
+        assertTrue(classInstance.hasReturnType(setIntAttributeName1, new Class[]{int.class}, void.class));
+    }
+
+    @ParameterizedTest
+    @MethodSource("constructorInputProvider")
+    public void fishClassHasCorrectToStringMethod(String colour, String gender, int age) throws Throwable {
```
Comment (Hedgemon4) [inline src/oop/object_independence/hard/q2/MainTest.java:265]: I think this should also be a dependent test

**Review by Hedgemon4 — CHANGES_REQUESTED** [review]

**Review by Hedgemon4 — APPROVED** [review]
